File: solve/cpp/spread.cc
#include "common.hh"

bool b[256];

void recurse(int i)
{
    if (i == n)
    {
        output();
    }
    else
    {
        for (int c = FIRST; c != LAST; ++c)
        {
            if (b[c])
            {
                a[i] = c;
                b[c] = false;
                recurse(i + 1);
                b[c] = true;
            }
        }       
    } 
}

int main(int argc, char *argv[])
{
	if (init(argc, argv, ORDERED) == 0)
    {
        fill(b, b + 256, true);
        recurse(0); 
        return final();
    }
    return 1;
}

File: solve/lua/spread.lua
#! /usr/bin/lua

require "common"

used = {}

function recurse(k)
	for i = 1, n do
		if not used[i] then
			a[k] = 96 + i
			if k == n then
				output()
			else
				used[k] = true
				recurse(k + 1)
				used[k] = false
			end
		end
	end
end

skeleton(ORDERED, function ()
	for i = 1, n do
		used[i] = false
	end 
	recurse(1)
end)
